migrate existing tutorial (#1329)
corda/corda-docs-portal: content/en/platform/corda/5.0/developing-applications/basic-cordapp/contract.md

This is what your code should look like now:

```kotlin
package com.r3.developers.apples.contracts

import net.corda.v5.ledger.utxo.Command

// Used to indicate the transaction's intent
interface AppleCommands : Command {
    class Issue : AppleCommands
    class Redeem : AppleCommands
}
  ```

### Add the `verify` Method

Returning to `AppleStampContract`, the `verify` method is automatically triggered when your transaction is executed.
It verifies that the transaction components are following the restrictions implemented inside the contract's `verify` method.

1. If you're using IntelliJ, you will see an error indicator under the class name and implementation. This indicates that the class is missing the required method. Hover over the class definition, then:
   * On macOS: press **Option** + **Enter**.
   * On Windows: press **Alt** + **Enter**.
2. Select **Implement methods > verify** from the dropdown menu.
   The `verify` method preceded by the `override` keyword appears.
3. Extract the command from the transaction.
4. Verify the intention of the transaction (Issue or Redeem) using an `if/else` or `when` block.

> 5. Use Kotlin’s `requires` method to include the contract’s verification rules for issuing:

Reviewer: ```suggestion
5. Use Kotlin’s `require` method to include the contract’s verification rules for issuing:
```

   ```kotlin
   val output = transaction.getOutputStates(AppleStamp::class.java).first()
   require(transaction.outputContractStates.size == 1) {
       "This transaction should only have one AppleStamp state as output"
   }
   require(output.stampDesc.isNotBlank()) {
       "The output AppleStamp state should have clear description of the type of redeemable goods"
   }
   ```
6. Similarly, add verification rules for redeeming:

{{< note >}}

> When the intention of the transaction is not recognized by the `verify` method, use `else` to throw an error.

Reviewer: this is unclear

                val output = transaction.getOutputStates(AppleStamp::class.java).first()
                require(transaction.outputContractStates.size == 1) {
                    "This transaction should only have one AppleStamp state as output"
                }
                require(output.stampDesc.isNotBlank()) {
                    "The output AppleStamp state should have clear description of the type of redeemable goods"
                }
            }
            is AppleCommands.Redeem -> {
                // Transaction verification will happen in BasketOfApplesContract
            }
            else -> {
                // Unrecognised Command type
                throw IllegalArgumentException("Incorrect type of AppleStamp commands: ${command::class.java.name}")
            }
        }
    }
}

```

## Create the `BasketOfApplesContract`

The `BasketOfApplesContract` has two intentions:

* Farmer Bob creates the basket of apples. This intention is expressed by the `PackBasket` command.
* Dave redeems the `BasketOfApples` state. This intention is expressed by the `Redeem` command, which you defined when creating `AppleStampContract`.

> The rules inside the `verify` method in the `requireThat` Corda DSL helper method are:

Reviewer: ```suggestion
The rules inside the `verify` method in the `require` Corda DSL helper method are:
```
?

  * The issuer of the `AppleStamp` should be the producing farm of this basket of apples.
  * The weight of the basket of apples must be greater than zero.

Using what you learned when writing `AppleStampContract`, implement the above in a new class. You will also need to
update the `AppleCommands` interface defined earlier to account for the new `PackBasket` command.

### Check Your Work

Once you have written the `BasketOfApplesContract`, check your code against the samples below.

Your `AppleCommands` file should look like this:

```kotlin
package com.r3.developers.apples.contracts

import net.corda.v5.ledger.utxo.Command

// Used to indicate the transaction's intent
interface AppleCommands : Command {
    class Issue : AppleCommands
    class Redeem : AppleCommands
    class PackBasket : AppleCommands
}
```

Your BasketOfApplesContract file should look like this:

```kotlin
package com.r3.developers.apples.contracts

import com.r3.developers.apples.states.AppleStamp
import com.r3.developers.apples.states.BasketOfApples
import net.corda.v5.ledger.utxo.Contract
import net.corda.v5.ledger.utxo.transaction.UtxoLedgerTransaction

> class BasketOfApplesContract : Contract{

Reviewer: ```suggestion
class BasketOfApplesContract : Contract {
```

    override fun verify(transaction: UtxoLedgerTransaction) {
        // Extract the command from the transaction
        when (val command = transaction.commands.first()) {

>             is AppleCommands.PackBasket -> {

Reviewer: no case for `Issue` command

Reviewer 2: @yarivus that's intentional, because that command is only valid in the context of `AppleStamp`, not `BasketOfApples`

Reviewer: I hope it won't confuse the CorDapp developers following the documentation ...
Maybe worth adding a comment/note on this ?

Reviewer 2: Added as a suggested change above.

Reviewer 3: Change committed.

---
date: '2023-05-03'
title: "Write Contracts"
version: 'Corda 5.0'
menu:
  corda5:
    identifier: corda5-develop-first-cordapp-contract
    parent: corda5-develop-first-cordapp
    weight: 3000
---

# Write Contracts

This tutorial guides you through writing the two contracts you need in your CorDapp: `AppleStampContract` and `BasketOfApplesContract`. You will link these contracts to the states that you created in the [Write States]({{< relref"./state.md" >}}) tutorial.

You will create these contracts in the `contracts/src/main/kotlin/com/r3/developers/apples/contracts` directory in this tutorial.

## Learning Objectives

Once you have completed this tutorial, you will know how to create and implement contracts in a CorDapp to restrict how your transaction flows are performed.

> 

Author: ```suggestion
```

1. If you're using IntelliJ, you will see an error indicator under the class name and implementation. This indicates that the class is missing the required method. Hover over the class definition, then:
   * On macOS: press **Option** + **Enter**.
   * On Windows: press **Alt** + **Enter**.
2. Select **Implement methods > verify** from the dropdown menu.
   The `verify` method preceded by the `override` keyword appears.
3. Extract the command from the transaction.
4. Verify the intention of the transaction (Issue or Redeem) using an `if/else` or `when` block.
5. Use Kotlin’s `require` method to include the contract’s verification rules for issuing:
   ```kotlin
   val output = transaction.getOutputStates(AppleStamp::class.java).first()
   require(transaction.outputContractStates.size == 1) {
       "This transaction should only have one AppleStamp state as output"
   }
   require(output.stampDesc.isNotBlank()) {
       "The output AppleStamp state should have clear description of the type of redeemable goods"
   }
   ```
6. Similarly, add verification rules for redeeming:

> {{< note >}}

Author: ```suggestion
   {{< note >}}
```

> When the intention of the transaction is not recognized by the `verify` method, use an `else` branch within your `if`/`when` block.

Author: ```suggestion
   When the intention of the transaction is not recognized by the `verify` method, use an `else` branch within your `if`/`when` block.
```

> {{< /note >}}

Author: ```suggestion
   {{< /note >}}
```

                val output = transaction.getOutputStates(AppleStamp::class.java).first()
                require(transaction.outputContractStates.size == 1) {
                    "This transaction should only have one AppleStamp state as output"
                }
                require(output.stampDesc.isNotBlank()) {
                    "The output AppleStamp state should have clear description of the type of redeemable goods"
                }
            }
            is AppleCommands.Redeem -> {
                val inputs = transaction.getInputStates(AppleStamp::class.java)
                require(inputs.size == 1) {
                    "This transaction should only have one AppleStamp state as input"
                }
                require(transaction.signatories.contains(inputs.first().holder)) {
                    "The holder of the input AppleStamp state must be a signatory to the transaction"
                }
            }
            else -> {
                // Unrecognised Command type
                throw IllegalArgumentException("Incorrect type of AppleStamp commands: ${command::class.java.name}")
            }
        }
    }
}

```

### Add Imports

> If you are using IntelliJ or another IDE, the IDE will automatically add the imports you need. IntelliJ indicates that an import is missing with red text.

Author: ```suggestion
If you are using IntelliJ or another IDE, the IDE automatically adds the imports you need. IntelliJ indicates that an import is missing with red text.
```
